File: models/deepseek_v32_exp/utils/compare.py
#!/usr/bin/env python3
# coding: utf-8
# Copyright (c) 2025-2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------
"""
"""
import os
import math
import logging
import torch
import numpy as np
from numpy.testing import assert_allclose
import pypto


def compare(t: torch.Tensor, t_ref: torch.Tensor, name, atol, rtol, max_error_ratio=0.005, max_error_count=10):
    """
    比较两个张量的差异，超过阈值时打印错误点并抛出断言错误
    Args:
        t: 待比较张量
        t_ref: 参考张量
        name: 张量名称（用于日志）
        atol: 绝对容差
        rtol: 相对容差
        max_error_ratio: 误差点占总元素数的最大比例
        max_error_count: 显示的最大误差点数量（同时也是误差点阈值的上限）
    """
    def check_is_nan_inf():
        # ========== 核心新增：检测t中的NaN和Inf并直接报错 ==========
        # 1. 检测NaN
        nan_mask = torch.isnan(t)
        nan_count = nan_mask.sum().item()

        # 2. 检测Inf（包含+Inf和-Inf）
        inf_mask = torch.isinf(t)
        inf_count = inf_mask.sum().item()

        # 若存在NaN或Inf，拼接错误信息并报错
        if nan_count > 0 or inf_count > 0:
            error_msg = f"\n========== 张量 {name} 检测到非法值（禁止存在NaN/Inf）=========="

            # 打印NaN信息
            if nan_count > 0:
                nan_positions = torch.nonzero(nan_mask, as_tuple=False)
                show_nan_count = min(nan_count, max_error_count)
                error_msg += f"\n- NaN数量：{nan_count}，前 {show_nan_count} 个位置："
                for i in range(show_nan_count):
                    pos_tuple = tuple(p.item() for p in nan_positions[i])
                    error_msg += f"\n  位置 {pos_tuple}"

            # 打印Inf信息（区分+Inf/-Inf）
            if inf_count > 0:
                inf_positions = torch.nonzero(inf_mask, as_tuple=False)
                show_inf_count = min(inf_count, max_error_count)
                error_msg += f"\n- Inf数量：{inf_count}，前 {show_inf_count} 个位置（值类型）："
                for i in range(show_inf_count):
                    pos = inf_positions[i]
                    pos_tuple = tuple(p.item() for p in pos)
                    inf_val = t[pos_tuple].item()
                    inf_type = "+Inf" if inf_val == float('inf') else "-Inf"
                    error_msg += f"\n  位置 {pos_tuple}：{inf_type}"
            error_msg += "\n" + "=" * 80 + "\n"

            # 抛出断言错误，终止函数执行
            assert False, error_msg

    # check 是否是nan 或 inf
    check_is_nan_inf()

    # 先验证张量的基本属性一致
    assert t.shape == t_ref.shape, f"张量形状不一致：t.shape={t.shape}, t_ref.shape={t_ref.shape}"
    assert t.dtype == t_ref.dtype, f"张量数据类型不一致：t.dtype={t.dtype}, t_ref.dtype={t_ref.dtype}"
    assert t.device == t_ref.device, f"张量设备不一致：t.device={t.device}, t_ref.device={t_ref.device}"

    # 计算误差点数量的阈值（取比例计算值和最大数量的较小值）
    error_count_threshold = round(max_error_ratio * t_ref.numel())

    # 计算误差掩码（超过阈值的位置为True）
    diff_abs = (t - t_ref).abs()
    tolerance = atol + rtol * t_ref.abs()
    diff_mask = diff_abs > tolerance
    error_count = diff_mask.sum().item()

    # 计算最大误差和其位置
    max_diff, flat_max_pos = torch.max(diff_abs.flatten(), dim=0)
    max_pos = torch.unravel_index(flat_max_pos, t.shape)
    max_pos = tuple(idx.item() for idx in max_pos)

    # 打印错误点的逻辑（如果有误差点）
    if error_count > 0:
        print(f"\n========== 张量 {name} 存在 {error_count} 个误差点（阈值：{error_count_threshold}）==========")

        # 获取所有误差点的位置
        error_positions = torch.nonzero(diff_mask, as_tuple=False)  # shape: [error_count, dims]

        # 限制显示的误差点数量（避免数据量过大）
        show_count = min(error_count, max_error_count)
        print(f"显示前 {show_count} 个误差点（位置 | 待比较值 | 参考值 | 绝对误差 | 允许阈值）：")

        # 遍历前N个误差点打印详细信息
        for i in range(show_count):
            pos = error_positions[i]

            # 转换为元组格式的位置（如 (0, 2, 3)）
            pos_tuple = tuple(p.item() for p in pos)

            # 获取对应位置的数值
            t_val = t[pos_tuple].item()
            t_ref_val = t_ref[pos_tuple].item()
            diff_val = diff_abs[pos_tuple].item()
            tol_val = tolerance[pos_tuple].item()

            # 格式化输出，保留足够小数位
            print(f"  位置 {pos_tuple}: {t_val:.8f} vs {t_ref_val:.8f} | 误差={diff_val:.8f} | 阈值={tol_val:.8f}")

        # 打印最大误差点
        print(f"\n最大误差点：位置 {max_pos} | 误差={max_diff.item():.8f} | 阈值={tolerance[max_pos].item():.8f}")
        print("=" * 80 + "\n")

    # 断言误差点数量不超过阈值
    assert error_count <= error_count_threshold, \
        (f"compare fail: {name}, max diff: {max_diff.item():.8f} at {max_pos}, "
         f"error_count: {error_count}, error_count_threshold: {error_count_threshold}")

File: framework/src/interface/operation/vector/compare.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

/*!
 * \file compare.cpp
 * \brief
 */

#include "binary.h"
#include "tensor_transformation.h"
#include "interface/utils/operator_tracer.h"
#include "passes/pass_utils/graph_utils.h"
#include "interface/utils/vector_error.h"

namespace npu::tile_fwk {

void TiledCompareOperationImpl(
    Function& function, const TileShape& tileShape, size_t cur, Input& input1, Input& input2,
    const LogicalTensorPtr& result, TileInfo& resultTileInfo, OpType operation, OutType mode)
{
    if (cur == result->shape.size()) {
        auto inputTile1 = input1.tensor.GetStorage()->View(function, input1.tileInfo.shape, input1.tileInfo.offset);
        auto inputTile2 = input2.tensor.GetStorage()->View(function, input2.tileInfo.shape, input2.tileInfo.offset);
        auto resultTile = result->View(function, resultTileInfo.shape, resultTileInfo.offset);

        LogicalTensorPtr convertedTile1 = inputTile1;
        LogicalTensorPtr convertedTile2 = inputTile2;

        if (inputTile1->Datatype() == DT_BF16) {
            convertedTile1 = std::make_shared<LogicalTensor>(function, DT_FP32, inputTile1->GetShape());
            Operation& castOp1 = function.AddOperation(Opcode::OP_CAST, {inputTile1}, {convertedTile1});
            castOp1.SetAttribute(OP_ATTR_PREFIX + "mode", CastMode::CAST_NONE);
            convertedTile2 = std::make_shared<LogicalTensor>(function, DT_FP32, inputTile2->GetShape());
            Operation& castOp2 = function.AddOperation(Opcode::OP_CAST, {inputTile2}, {convertedTile2});
            castOp2.SetAttribute(OP_ATTR_PREFIX + "mode", CastMode::CAST_NONE);
        }

        const int64_t COUNT_MODE_SIZE = 4096;
        size_t element_size = BytesOf(input1.tensor.GetDataType());
        if (inputTile1->Datatype() == DT_BF16) {
            element_size = BytesOf(DT_FP32);
        }
        ASSERT(VectorErrorCode::ERR_RUNTIME_LOGIC, element_size != 0) << "Element size cannot be zero.";
        int64_t elements_per_chunk = COUNT_MODE_SIZE / element_size;
        int64_t vcmp_bits_size = (elements_per_chunk + 7) / 8;

        const size_t ALIGN_SIZE = 32;

        size_t vcmpBitResult_size = ((vcmp_bits_size + ALIGN_SIZE - 1) / ALIGN_SIZE) * ALIGN_SIZE;
        size_t array_size = elements_per_chunk * element_size;
        size_t aligned_array_size = ((array_size + ALIGN_SIZE - 1) / ALIGN_SIZE) * ALIGN_SIZE;

        size_t total_bytes = vcmpBitResult_size + 3 * aligned_array_size + ALIGN_SIZE * 2;
        std::vector<int64_t> tmp_shape({static_cast<int64_t>(total_bytes)});
        auto tmp_tensor = std::make_shared<LogicalTensor>(function, DT_UINT8, tmp_shape);

        auto& op = function.AddOperation(Opcode::OP_CMP, {convertedTile1, convertedTile2}, {resultTile, tmp_tensor});
        std::vector<bool> dimMap({true, true});
        op.SetAttr(OpAttributeKey::rowPad, dimMap);

        op.SetAttribute(OP_ATTR_PREFIX + "cmp_operation", static_cast<int64_t>(operation));
        op.SetAttribute(OP_ATTR_PREFIX + "cmp_mode", static_cast<int64_t>(mode));
        return;
    }

    auto& vecTile = tileShape.GetVecTile();
    int64_t step = vecTile[cur];

    if (mode == OutType::BIT && cur == result->shape.size() - 1) {
        step = vecTile[cur] / NUM_VALUE_8;
        if (step < 1)
            step = 1;

        int64_t actualInputStep = step * NUM_VALUE_8;

        for (int i = 0; i < result->shape[cur]; i += step) {
            resultTileInfo.offset[cur] = i;
            resultTileInfo.shape[cur] = std::min(result->shape[cur] - i, step);

            input1.tileInfo.offset[cur] = (i * NUM_VALUE_8) % input1.tensor.GetShape()[cur];
            input1.tileInfo.shape[cur] =
                std::min(input1.tensor.GetShape()[cur] - input1.tileInfo.offset[cur], actualInputStep);
            input2.tileInfo.offset[cur] = (i * NUM_VALUE_8) % input2.tensor.GetShape()[cur];
            input2.tileInfo.shape[cur] =
                std::min(input2.tensor.GetShape()[cur] - input2.tileInfo.offset[cur], actualInputStep);

            TiledCompareOperationImpl(
                function, tileShape, cur + 1, input1, input2, result, resultTileInfo, operation, mode);
        }
    } else {
        for (int i = 0; i < result->shape[cur]; i += step) {
            resultTileInfo.offset[cur] = i;
            resultTileInfo.shape[cur] = std::min(result->shape[cur] - i, step);

            input1.tileInfo.offset[cur] = i % input1.tensor.GetShape()[cur];
            input1.tileInfo.shape[cur] = std::min(input1.tensor.GetShape()[cur] - input1.tileInfo.offset[cur], step);

            input2.tileInfo.offset[cur] = i % input2.tensor.GetShape()[cur];
            input2.tileInfo.shape[cur] = std::min(input2.tensor.GetShape()[cur] - input2.tileInfo.offset[cur], step);

            TiledCompareOperationImpl(
                function, tileShape, cur + 1, input1, input2, result, resultTileInfo, operation, mode);
        }
    }
}

void TiledCompareOperation(
    Function& function, const TileShape& tileShape, LogicalTensorPtr operand1, LogicalTensorPtr operand2,
    const LogicalTensorPtr& result, OpType operation, OutType mode)
{
    auto broadcastOperand = [&](LogicalTensorPtr& operand, LogicalTensorPtr& other) {
        auto dstShape = result->shape;
        if (mode == OutType::BIT) {
            dstShape[dstShape.size() - 1] *= NUM_VALUE_8;
        }
        if (operand->shape == dstShape) {
            return;
        }
        auto expanded = std::make_shared<LogicalTensor>(function, operand->Datatype(), dstShape);
        Expand(function, tileShape, operand, {other}, expanded);
        operand = expanded;
    };
    broadcastOperand(operand1, operand2);
    broadcastOperand(operand2, operand1);

    TileInfo tileInfo1(result->shape.size(), result->offset.size());
    TileInfo tileInfo2(result->shape.size(), result->offset.size());
    TileInfo resultTileInfo(result->shape.size(), result->offset.size());
    auto input1 = Input{operand1, tileInfo1};
    auto input2 = Input{operand2, tileInfo2};

    TiledCompareOperationImpl(function, tileShape, 0, input1, input2, result, resultTileInfo, operation, mode);
}

LogicalTensorPtr TensorCompareOperation(
    Function& function, const Tensor& self, const Tensor& other, OpType operation, OutType mode)
{
    auto operandT1 = self.GetStorage();
    auto operandT2 = other.GetStorage();
    if (operandT1->shape.size() != operandT2->shape.size()) {
        std::vector<int> broadCastShape = GetBroadCastShape(operandT1, operandT2);
        operandT1 = BinaryOperationBroadCast(operandT1, broadCastShape);
        operandT2 = BinaryOperationBroadCast(operandT2, broadCastShape);
    }
    std::vector<SymbolicScalar> resultValidShape;
    std::vector<int64_t> resultShape = BinaryOperationResultShape(operandT1, operandT2);
    if (!operandT1->GetDynValidShape().empty() && !operandT2->GetDynValidShape().empty()) {
        for (size_t i = 0; i < resultShape.size(); ++i) {
            if (resultShape[i] == operandT1->shape[i]) {
                resultValidShape.push_back(operandT1->GetDynValidShape()[i]);
            } else {
                resultValidShape.push_back(operandT2->GetDynValidShape()[i]);
            }
        }
    }
    auto resultType = DT_BOOL;
    if (mode == OutType::BIT) {
        resultType = DT_UINT8;
        ASSERT(VectorErrorCode::ERR_CONFIG_ALIGNMENT, resultShape.empty() || resultShape.back() % NUM_VALUE_8 == 0)
            << "Last dimension must be divisible by 8 in BIT mode";
        if (!resultShape.empty()) {
            resultShape.back() /= NUM_VALUE_8;
            if (!resultValidShape.empty()) {
                resultValidShape.back() = resultValidShape.back() / NUM_VALUE_8;
            }
        }
    }
    auto result = std::make_shared<LogicalTensor>(function, resultType, resultShape, resultValidShape);
    auto& op = function.AddOperation(Opcode::OP_CMP, {operandT1, operandT2}, {result});
    std::vector<bool> dimMap({true, true});
    op.SetAttr(OpAttributeKey::rowPad, dimMap);
    op.SetAttribute(OP_ATTR_PREFIX + "cmp_operation", static_cast<int64_t>(operation));
    op.SetAttribute(OP_ATTR_PREFIX + "cmp_mode", static_cast<int64_t>(mode));
    return result;
}

LogicalTensorPtr TensorCompareOperationScalar(
    Function& function, const Tensor& operand1, const Element& value, OpType operation, OutType mode)
{
    DECLARE_TRACER();
    auto operandT1 = operand1.GetStorage();
    std::vector<int64_t> resultShape = operandT1->shape;
    std::vector<SymbolicScalar> resultValidShape = operandT1->GetDynValidShape();
    DataType resultType = DT_BOOL;
    if (mode == OutType::BIT) {
        resultType = DT_UINT8;
        if (!resultShape.empty()) {
            int64_t lastDim = resultShape.back();
            ASSERT(VectorErrorCode::ERR_CONFIG_ALIGNMENT, lastDim % NUM_VALUE_8 == 0)
                << "Last dimension must be divisible by 8 in BIT mode";
            resultShape.back() = lastDim / NUM_VALUE_8;
            if (!resultValidShape.empty()) {
                auto& lastSymDim = resultValidShape.back();
                resultValidShape.back() = lastSymDim / NUM_VALUE_8;
            }
        }
    }
    auto result = std::make_shared<LogicalTensor>(function, resultType, resultShape, resultValidShape);
    auto& op = function.AddOperation(Opcode::OP_CMPS, {operandT1}, {result});
    std::vector<bool> dimMap({true});
    op.SetAttr(OpAttributeKey::rowPad, dimMap);
    op.SetAttribute(OpAttributeKey::scalar, value);
    op.SetAttribute(OP_ATTR_PREFIX + "cmp_operation", static_cast<int64_t>(operation));
    op.SetAttribute(OP_ATTR_PREFIX + "cmp_mode", static_cast<int64_t>(mode));

    return result;
}

LogicalTensorPtr TensorCompareOperationScalar(
    Function& function, const Element& value, const Tensor& operand1, OpType operation, OutType mode)
{
    switch (operation) {
        case OpType::LT:
            operation = OpType::GT;
            break;
        case OpType::GT:
            operation = OpType::LT;
            break;
        case OpType::LE:
            operation = OpType::GE;
            break;
        case OpType::GE:
            operation = OpType::LE;
            break;
        default:
            break;
    }
    Element converted_value = value;
    if (value.GetDataType() == DataType::DT_BF16) {
        double val = value.GetFloatData();
        converted_value = Element(DataType::DT_FP32, val);
    }

    return TensorCompareOperationScalar(function, operand1, converted_value, operation, mode);
}

void TiledCmpsOperationImpl(
    Function& function, const TileShape& tileShape, size_t cur, Input& input, const Element& scalar,
    const LogicalTensorPtr& result, TileInfo& resultTileInfo, OpType operation, OutType mode)
{
    if (cur == result->shape.size()) {
        auto inputTile = input.tensor.GetStorage()->View(function, input.tileInfo.shape, input.tileInfo.offset);
        auto resultTile = result->View(function, resultTileInfo.shape, resultTileInfo.offset);

        LogicalTensorPtr convertedTile = inputTile;
        if (inputTile->Datatype() == DT_BF16) {
            convertedTile = std::make_shared<LogicalTensor>(function, DT_FP32, inputTile->GetShape());
            Operation& castOp = function.AddOperation(Opcode::OP_CAST, {inputTile}, {convertedTile});
            castOp.SetAttribute(OP_ATTR_PREFIX + "mode", CastMode::CAST_NONE);
        }

        const int64_t COUNT_MODE_SIZE = 4096;
        size_t element_size = BytesOf(input.tensor.GetDataType());
        if (inputTile->Datatype() == DT_BF16) {
            element_size = BytesOf(DT_FP32);
        }
        ASSERT(VectorErrorCode::ERR_RUNTIME_LOGIC, element_size != 0) << "Element size cannot be zero.";
        int64_t elements_per_chunk = COUNT_MODE_SIZE / element_size;
        int64_t vcmp_bits_size = (elements_per_chunk + 8 - 1) / 8;

        const size_t ALIGN_SIZE = 32;

        size_t vcmpBitResult_size = ((vcmp_bits_size + ALIGN_SIZE - 1) / ALIGN_SIZE) * ALIGN_SIZE;
        size_t array_size = elements_per_chunk * element_size;
        size_t aligned_array_size = ((array_size + ALIGN_SIZE - 1) / ALIGN_SIZE) * ALIGN_SIZE;

        size_t total_bytes = vcmpBitResult_size + 3 * aligned_array_size + ALIGN_SIZE;
        std::vector<int64_t> tmp_shape({static_cast<int64_t>(total_bytes)});
        auto tmp_tensor = std::make_shared<LogicalTensor>(function, DT_UINT8, tmp_shape);

        auto& op = function.AddOperation(Opcode::OP_CMPS, {convertedTile}, {resultTile, tmp_tensor});
        std::vector<bool> dimMap({true});
        op.SetAttr(OpAttributeKey::rowPad, dimMap);

        op.SetAttribute(OP_ATTR_PREFIX + "cmp_operation", static_cast<int64_t>(operation));
        op.SetAttribute(OP_ATTR_PREFIX + "cmp_mode", static_cast<int64_t>(mode));
        op.SetAttribute(OpAttributeKey::scalar, scalar);
        return;
    }

    auto& vecTile = tileShape.GetVecTile();
    int64_t step = vecTile[cur];

    if (mode == OutType::BIT && cur == result->shape.size() - 1) {
        step = vecTile[cur] / NUM_VALUE_8;
        if (step < 1)
            step = 1;

        int64_t actualInputStep = step * NUM_VALUE_8;

        for (int i = 0; i < result->shape[cur]; i += step) {
            resultTileInfo.offset[cur] = i;
            resultTileInfo.shape[cur] = std::min(result->shape[cur] - i, step);

            input.tileInfo.offset[cur] = (i * NUM_VALUE_8) % input.tensor.GetShape()[cur];
            input.tileInfo.shape[cur] =
                std::min(input.tensor.GetShape()[cur] - input.tileInfo.offset[cur], actualInputStep);

            TiledCmpsOperationImpl(
                function, tileShape, cur + 1, input, scalar, result, resultTileInfo, operation, mode);
        }
    } else {
        for (int i = 0; i < result->shape[cur]; i += step) {
            resultTileInfo.offset[cur] = i;
            resultTileInfo.shape[cur] = std::min(result->shape[cur] - i, step);

            input.tileInfo.offset[cur] = i % input.tensor.GetShape()[cur];
            input.tileInfo.shape[cur] = std::min(input.tensor.GetShape()[cur] - input.tileInfo.offset[cur], step);

            TiledCmpsOperationImpl(
                function, tileShape, cur + 1, input, scalar, result, resultTileInfo, operation, mode);
        }
    }
}

void TiledCmpsOperation(
    Function& function, const TileShape& tileShape, LogicalTensorPtr operand, const Element& scalar,
    const LogicalTensorPtr& result, OpType operation, OutType mode)
{
    TileInfo tileInfo(result->shape.size(), result->offset.size());
    TileInfo resultTileInfo(result->shape.size(), result->offset.size());
    auto input = Input{operand, tileInfo};
    TiledCmpsOperationImpl(function, tileShape, 0, input, scalar, result, resultTileInfo, operation, mode);
}

Tensor Compare(const Tensor& self, const Tensor& other, OpType op, OutType mode)
{
    DECLARE_TRACER();
    RETURN_CALL(CompareOperation, *Program::GetInstance().GetCurrentFunction(), self, other, op, mode);
}

Tensor Compare(const Tensor& self, const Element& other, OpType op, OutType mode)
{
    DECLARE_TRACER();
    RETURN_CALL(CompareOperationScalar, *Program::GetInstance().GetCurrentFunction(), self, other, op, mode);
}

Tensor Compare(const Element& self, const Tensor& other, OpType op, OutType mode)
{
    DECLARE_TRACER();
    RETURN_CALL(CompareOperationScalar, *Program::GetInstance().GetCurrentFunction(), self, other, op, mode);
}

void CompareOperationTileFunc(
    Function& function, const TileShape& tileShape, const std::vector<LogicalTensorPtr>& iOperand,
    const std::vector<LogicalTensorPtr>& oOperand, const Operation& op)
{
    BinaryOperationOperandCheck(iOperand, oOperand);
    auto operation = static_cast<OpType>(op.GetIntAttribute(OP_ATTR_PREFIX + "cmp_operation"));
    auto mode = static_cast<OutType>(op.GetIntAttribute(OP_ATTR_PREFIX + "cmp_mode"));
    TiledCompareOperation(function, tileShape, iOperand[0], iOperand[1], oOperand[0], operation, mode);
}

void CmpsOperationTileFunc(
    Function& function, const TileShape& tileShape, const std::vector<LogicalTensorPtr>& iOperand,
    const std::vector<LogicalTensorPtr>& oOperand, const Operation& op)
{
    auto operation = static_cast<OpType>(op.GetIntAttribute(OP_ATTR_PREFIX + "cmp_operation"));
    auto mode = static_cast<OutType>(op.GetIntAttribute(OP_ATTR_PREFIX + "cmp_mode"));
    TiledCmpsOperation(
        function, tileShape, iOperand[0], op.GetElementAttribute(OpAttributeKey::scalar), oOperand[0], operation, mode);
}

REGISTER_OPERATION_TILED_FUNC(OP_CMP, Opcode::OP_CMP, CompareOperationTileFunc);
REGISTER_OPERATION_TILED_FUNC(OP_CMPS, Opcode::OP_CMPS, CmpsOperationTileFunc);

} // namespace npu::tile_fwk
